cmd/gardener-controller-manager/app/gardener_controller_manager.go

```diff
@@ -302,49 +303,47 @@ func NewGardener(config *componentconfig.ControllerManagerConfiguration) (*Garde
 }
 
 // Run runs the Gardener. This should never exit.
-func (g *Gardener) Run(stopCh chan struct{}) error {
+func (g *Gardener) Run(ctx context.Context, cancel context.CancelFunc) error {
+	leaderElectionCtx, leaderElectionCancel := context.WithCancel(context.Background())
+
 	// Prepare a reusable run function.
-	run := func(stop <-chan struct{}) {
-		go g.startControllers(stopCh)
-		<-stop
-		select {
-		case <-stopCh:
-			// can only happen if stopCh is already closed because it's never written to it
-		default:
-			close(stopCh)
-		}
+	run := func(ctx context.Context) {
+		g.startControllers(ctx)
 	}
 
 	// Start HTTP server
-	go server.Serve(g.K8sGardenClient, g.K8sGardenInformers, g.Config.Server, stopCh)
+	go server.Serve(ctx, g.K8sGardenClient, g.K8sGardenInformers, g.Config.Server)
 	handlers.UpdateHealth(true)
 
 	// If leader election is enabled, run via LeaderElector until done and exit.
 	if g.LeaderElection != nil {
 		g.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
-			OnStartedLeading: func(stop <-chan struct{}) {
+			OnStartedLeading: func(_ context.Context) {
 				g.Logger.Info("Acquired leadership, starting controllers.")
-				run(stop)
+				run(ctx)
+				<-ctx.Done()
```

Comment:
`run(ctx)` already waits for the channel to be closed. Do we really have to it here again?